### ISSUE TITLE
Should Dropout time the rate when feed forward propagation? 

### ISSUE THREAD

**Issue by pengsun** [vlfeat/matconvnet#76]
In `vl_simplenn.m`, the codes involving dropout look like
``` Matlab
    case 'dropout'
     if opts.disableDropout
        res(i+1).x = res(i).x ;
      elseif opts.freezeDropout
        [res(i+1).x, res(i+1).aux] = vl_nndropout(res(i).x, 'rate', l.rate, 'mask', res(i+1).aux) ;
      else
        [res(i+1).x, res(i+1).aux] = vl_nndropout(res(i).x, 'rate', l.rate) ;
      end
```

Should it be 
``` Matlab
res(i+1).x = l.rate * res(i).x ;
```
according to what is specified in the JMLR dropout paper?

**Comment by felixachilles** [comment]
We had the same impression but did not test on a bigger dataset yet. Could
you share your experiences?

**Comment by lenck** [comment]
Hi, when you look at the implementation of `vl_nndropout`, the activations are scaled by a factor `1/(1-rate)` which makes sense to me, as with higher dropout rate you zero-out more activations, therefore in order to keep similar average you need to increase the activation's magnitude... 
In the JMLR article it seems that `r` is sample from the Bernoulli distribution with a rate `p`. I haven't found directly the authors to say what scaling do they use when they switch-off the dropout. However in section 5.2 they give a hint how to approach it when one wants to use dropout for networks pre-trained with unsepervised pre-training:
```
The weights obtained from pretraining should be scaled up by a factor of 1/p. 
This makes sure that for each unit, the expected output from it under random 
dropout will be the same as the output during pretraining.
```
Which, in effect is the same approach as we do, only that we multiply the activations during training so that when dropout is disabled the mean remains the same. Sort of black magic, really...

**Comment by vedaldi** [comment]
Hi, yes the is the correct interpretation. It is done in this way so that there is no need to rescale the learned filters. Simply get rid of the dropout layer(s) in testing.

**Comment by pengsun** [comment]
Hi, thanks so much for the detailed explanations. I didn't carefully read the codes, sorry for that :)

**Comment by YauPH** [comment]
In the commet of vl_nndropout.m, rate is said to be the probability that a variable is retained. While in the implementation, the activation in backmode is scaled by 1/(1-rate), it seems that the rate is the probability a variable to be zeroed.  I am a bit confused, which is right?
